### DISCUSSION TITLE
Feature f0024 - rxar and rxln

### DISCUSSION THREAD

**Discussion by mgrossmann** [adesutherland/CREXX#201] [Progress Updates]
Hello all.

I would be interested in your thoughts on this topic.
Since I lack a few weeks of knowledge, I'd like to catch up a bit here.

I am not quite clear on the difference between `rxar` and `rxln`. Under `rxar` I imagine a tool that behaves similar to the Java `jar`tool. We will put all assembled rexx binaries together in an archive and have to define the entry point somehow. (Similar to the Java `public static void main()` method. 

A `rxln` tool would do the same, but not as a "ZIP" file but as a new rexx binary, right?

What do you say?

Greetings Mike

**Comment by adesutherland** [comment]
I think so

- rxar is like ar (https://linux.die.net/man/1/ar) - except we only care about creating an archive. Basically we take each rxbin member and "concat" them together so that the rxln can read each rxbin member (rather than specifying each rxbin file separately). Very much like mainframe library files? The format just needs the reader to know when the rxbin member ends and the next starts. Maybe rxar also needs a way of listing members? Not sure about using zip as we will need to include an unzip library in the rxvm(?)

- rxln is the actual linker - and does/replaces what the code at the start of rxvm does, and yes needs to create a rxbin file which contains all (and only) the used members. The rxbin file format will need extending to handle the multiple (rather than 1) code and constant block if that makes sense?

**Reply by mgrossmann** [reply]
what's about starting with the linker? (because in the end the linker and the archiver are doing pretty much the same.)
I would make the linker using a directory to search for the needed rxbin files. Later we could add the "library" functionality.

What do you think?

**Reply by adesutherland** [reply]
Sounds good :-)

**Comment by mgrossmann** [comment]
Hello ...

what do you think about following interface to `rxar` ?

```
C:\Work\repositores\CREXX\cmake-build-debug\archiver\rxar.exe
No valid option specified.
cREXX Archiver vX.X.X
Usage   : rxar [options] library-name rxbin-file...
Options :
      --license   Display Copyright & License information.
  -h  --help      Display this usage information.
  -a  --add       Add binary to library.
  -l  --list      List content of library.
  -d  --delete    Delete binary from library.
  -v  --verbose   Enable verbose output.
```


Best, MikeG

**Reply by rvjansen** [reply]
This looks nice! I also thought about your initial question, although I have to reproduce it from memory as I cannot find it anywhere now. Apart from linking the .rxbin modules on a run from rxvm, it would also be nice to have a way to statically link the interpreter with a program and its functions, so that it could be used by another person (or ourselves, depending on the program). ooRexx can 'pre-tokenize', and with OS/2 this version was stored in an OS/2 Resource Fork, so with unchanged source, the program never was re-tokenized. Resource Forks on OS/2 and Mac are history now, but it seems useful until we can do full native compiles.

**Reply by rvjansen** [reply]
And if I may add one tiny thing, it would also safeguard against a program not running (dumping core) on a slightly newer version of the rxvm, might be good for utilities and such.

**Comment by adesutherland** [comment]
Looks great!

**Comment by mgrossmann** [comment]
Small update.

Kind of wildcard support.

Next on my list, filter out duplicate entries.

```
>>>  rxar --verbose --add mylib ~/git/CREXX/lib/rxfns/*/*.rxbin *.rxbin ./tstsubstr
Creating library './mylib.rxlib'. 
Library './mylib.rxlib' will be created with following binaries:
		- '/Users/mike/git/CREXX/lib/rxfns/rxas/substr.rxbin' 
		- '/Users/mike/git/CREXX/lib/rxfns/tests_functional/tstsabstr.rxbin' 
		- '/Users/mike/git/CREXX/lib/rxfns/tests_functional/tstsubstr.rxbin' 
		- './tstsabstr.rxbin' 
		- './tstsubstr.rxbin' 
		- './tstsubstr.rxbin' 
```

**Comment by mgrossmann** [comment]
Hi all,

I have some new questions. :)

```
>>> rxar.exe --verbose --add mylib *.rxbin
Adding to existing library '.\mylib.rxlib'.
Library '.\mylib.rxlib' will be created with following binaries:
                - '.\bla.rxbin'
                  Exposed procedures:
                        - 'test.pos'
                        - 'test.pos2'
                        - 'test.pos3'
                  Imported procedures:
                - '.\test.rxbin'
                  Exposed procedures:
                  Imported procedures:
                        - 'test.pos'
                        - 'test.pos2'
                        - 'test.pos3'
                - '.\testpos.rxbin'
                  Exposed procedures:
                  Imported procedures:
                        - 'global.pos'
                        - 'global.lastpos'
                        - 'global.copies'
```

@adesutherland @rvjansen @Peter-Jacob  My thoughts and questions:

Because I would like to include the names of exported procedures in the generated RXLIB file (and maybe of the imported, too):
 
1) In the procedure names, the first part is a kind of a namespace? So we can have multiple versions of a procedure?
2) What should I do in the archiver with nonunique procedure names? e.g.  Two rxbin files are exporting the exact same procedure, by name? If the archiver is a real archiver only, then I would put both rxbin files in the archive. But if the archiver a kind of prelinker, then we should handle such problems. But this will bring new problems. See question 3.
3) As one can see, a single rxbin file can export multiple procedures. What to do, when a rxbin file A, is exporting more then one procedure. e.g.  Proc1, Proc2 and Proc3. Next rxbin file B, also exporting multiple procedures. e.g. Proc3, Proc4 and Proc5. How do we handle Proc 3?
4) If the archiver is more like a prilinker, maybe should add an option to only put one specifiy namespace in a single archive / library.

I think there are more problems like 2,3 and 4.

Best, Mike

**Reply by mgrossmann** [reply]
This is why you @adesutherland wrote:
> The rxbin file format will need extending to handle the multiple (rather than 1) code and constant block if that makes sense?

right?

But this must be handled earlier in the assembler, isn't it? When writing the rxbin file, I simply write down the bin_space structure. If we had our own area in bin_space for every single exposed procedure, that would be nice.

**Reply by mgrossmann** [reply]
One more finding:

When using the disassembler on a rxbin file with multiple exposed procedures, the disassembler is dying. e.g. _datei

**Reply by adesutherland** [reply]
Firstly - apologies for being quiet for a bit - I have been struggling with some old shell spawning code that is needed for address. My port from C++ to C was imperfect (!) and the code is complex (lots of threads to handle differing in/out pipes/pseudo terminal control - lots of race conditions). Anyway I made a breakthrough yesterday - and "soon" will have the linux version working, windows does work, then macos). Why bother? Because it will give cREXX ADDRESS "cms pipe" like features, and more besides (like calling a REXX procedure each time an ADDRESS command needs an input string or produces output. But anyway - getting the old code working has proved more frustrating than expected!!

Anyway I think rxar should just act as an archiver ... so quite simplistic. It does not check for duplicate (or potentially missing) symbols - it kind of just plonks everything in one file. E.g. kind of like the program[mod_index] structure in the interpreter.

Now when we come to the linker you are right - we need to agree rules about duplicate symbols. At the moment the "linking" code in the interpreter (from about line 180) errors if there is a duplicate or missing symbol. We might agree that if reading an archive file the linker just ignores duplicates and only uses the first version - a common approach but less safe for finding bugs, so not sure??

>  disassembler on a rxbin file with multiple exposed procedures, the disassembler is dying. e.g. _datei

Oh no! - we should raise an issue to fix ...

> bin_space for every single exposed procedure

I think that this is in general impossible - and also not needed. We have a rxbin structure per source file, and that can continue even when we have packed them together in one archive. Even the linker will be obliged to pull in the whole rxfile structure if it needs just one of many exposed procedures because it wont know if one exposed procedure is calling other procedures in the same rxbin file, or what constants are needed, and this will be even more impossible to know once we have classes, polymorphism etc. So very very hard and no benefit - treat the rxbin as the "code quantum" :-)

> 1. In the procedure names, the first part is a kind of a namespace? So we can have multiple versions of a procedure?

Exactly - at the linker level it is just a name with some dots - but for the compiler the structure of the name tells it what the class is and so on. Also (to be designed) the name may give information about the expected arguments (this may be done in the constant pool however). So the rule is, for linker it is just a unique string of characters. For the compiler it has other meanings - and this all may need tweaking in the future!

> 2,3,4 

I think my preamble covered these - it's just a braindead archiver! :-)

The linker will be harder as we make it more sophisticated.

**Reply by mgrossmann** [reply]
Hey @adesutherland 

thanks for your reply. 

With all your input, the archiver will be really simple. (braindead) :)
Creating the archive is mostly done.  Will send you all an update this weekend. 


Mike

**Comment by mgrossmann** [comment]
Hey folks,

currently I could spend some time in CREXX. What is the current state of the archiver and the linker?
Adrian did you put any work into it? Or can I use my current state to "restart"?

Best Mike

**Reply by rvjansen** [reply]
Hi Mike, that is great news. In the meantime, I implemented my own request (#205) for your branch, which is the flexible extension on the filename for .rexx files, including the possibility to just specify .rexx (e.g. with tab completion) without that leading to an error, as stated in https://github.com/adesutherland/CREXX/issues/205 . If someone could check that and merge feature/f0039 into develop, we have no chance of that interfering with your activity. (And I can merge that into feature/f0038 then).

**Reply by mgrossmann** [reply]
@rvjansen is this something that I have to worry about? Don't think so, isn't. So I will simply rebase my f0024 if your f0039 is merged into develop.

**Reply by rvjansen** [reply]
@mgrossmann nothing to worry about, just fyi.

**Comment by adesutherland** [comment]
@mgrossmann What do you think about having one file type - rxbin - which can have either one or many modules in it, rather than having a separate rxlib file type. To do this we would add additional fields to rxbin (like your signature fields, etc.).

The assembler would generate a rxlib file with one module, the archiver tool could move modules around etc. as per the functionality you have developed.

If this makes sense we just need to agree the rxbin structure - I can update the assembler and rxvm, you can update the archiver tool.

What do you think - and what would be your suggested rxbin format?

Adrian

**Comment by mgrossmann** [comment]
@adesutherland  I‘m fine with this. Please give me another week of time. I would like to
think about the structure, again. It‘s my current implementation vs the
original ar implementation.

Nevertheless, I would be interested to know what the background to this idea is.

Mike

**Reply by adesutherland** [reply]
> I would be interested to know what the background to this idea is.

It is just to simplify/reduce the number of file types - and realising that the assembler can as easily create an archive file with one rxbin file inside as it can create the rxbin file. Also allows the work to be split a bit.
